src/gc/impl/conservative/gc.d

```diff
@@ -2060,35 +2062,40 @@ struct Gcx
                     if(pool.nointerior.nbits && pool.nointerior.test(biti))
                         continue;
 
-                    if (!pool.mark.set(biti) && !pool.noscan.test(biti)) {
-                        stack[stackPos++] = ScanRange(base, base + pool.bPageOffsets[pn] * PAGESIZE);
-                        if (stackPos == stack.length)
-                            break;
+                    if (!pool.mark.set(biti) && !pool.noscan.test(biti))
+                    {
+                        top = base + pool.bPageOffsets[pn] * PAGESIZE;
+                        goto LaddRange;
                     }
                 }
                 else
                 {
                     // Don't mark bits in B_FREE pages
                     assert(bin == B_FREE);
+                }
+                continue;
+
+            LaddRange:
+                if (stackPos < stack.length)
+                {
+                    stack[stackPos++] = ScanRange(base, top);
                     continue;
                 }
+                if (p1 + 1 < p2) // *p1 already scanned
+                    toscan.push(ScanRange(p1 + 1, p2));
+                // reverse order for depth-first-order traversal
+                foreach_reverse (ref rng; stack)
+                    toscan.push(rng);
+                stackPos = 0;
+                // continue with last stack entry
+                p1 = cast(void**)base;
+                p2 = cast(void**)top;
+                goto LnextBody; // skip increment and check
```

Comment:
That's a second loop around/inside the loop, can we keep reusing the Lagain part for that?
Maybe by hoisting the `next` variable and setting it here `next = ScanRange(base, top)`, then jumping to the `p1 = cast(void**)next.pbot;` part.
Seems a bit cleaner to me and the case of the local stack overflowing is rare enough to not worry about 2 cycles or so.